### ISSUE TITLE
Fails on Node 0.12 and iojs

### ISSUE THREAD

**Issue by sindresorhus** [sindresorhus/to-fast-properties#2]
See: https://github.com/sindresorhus/to-fast-properties/pull/1

Thoughts on how to make it work again?

// @eush77 @d10

**Comment by sindresorhus** [comment]
This seems to fix the tests, but I have no idea why.

Inspired by https://code.google.com/p/v8/source/browse/branches/bleeding_edge/test/mjsunit/fast-prototype.js#52

``` diff
  module.exports = function toFastProperties(obj) {
    /*jshint -W027*/
    function f() {}
    f.prototype = obj;
-   return f;
+   new f();
+   return;
    eval(obj);
 };
```

**Comment by bnjmnt4n** [comment]
Not too sure, actually.

**Comment by eush77** [comment]
I finally checked out some V8 versions and done some investigation.

In 3.14.5.10 (the version used in Node 0.10) function prototypes are being set through [JSFunction::SetPrototype](https://github.com/v8/v8-git-mirror/blob/3.14.5.10/src/objects.cc#L7779) and then [JSFunction::SetInstancePrototype](https://github.com/v8/v8-git-mirror/blob/3.14.5.10/src/objects.cc#L7731) which contains these lines:

``` c++
  // First some logic for the map of the prototype to make sure it is in fast
  // mode.
  if (value->IsJSObject()) {
    MaybeObject* ok = JSObject::cast(value)->OptimizeAsPrototype();
    if (ok->IsFailure()) return ok;
  }
```

`OptimizeAsPrototype` is where all the conversion happens.

In 4.1.0.21 (latest iojs) [JSFunction::SetPrototype](https://github.com/v8/v8-git-mirror/blob/4.1.0.21/src/objects.cc#L10079) calls [JSFunction::SetInstancePrototype](https://github.com/v8/v8-git-mirror/blob/4.1.0.21/src/objects.cc#L10026) (same story), but those lines are gone. V8 no longer optimizes prototype objects this way.

Instead, on newer V8 `new f()` triggers this call chain that ultimately leads to `OptimizeAsPrototype`:
- `Factory::NewJSObject` in factory.cc
- `JSFunction::EnsureHasInitialMap` in objects.cc
- `JSFunction::SetInitialMap` in objects.cc
- `Map::SetPrototype` in objects.cc
- `JSObject::OptimizeAsPrototype` in objects.cc

What is quite interesting though is that the most latest V8 [brings that `OptimizeAsPrototype` call in `SetInstancePrototype` back](https://github.com/v8/v8-git-mirror/blob/e522cb6e3fa139682967ff0669b6923f08cb9b61/src/objects.cc#L10116-10117), which means that `toFastProperties` as it is now might work on some future version of iojs/Node.

After all, I think this change does actually fix the behavior on Node 0.12 and iojs and is worth putting in.

**Comment by sindresorhus** [comment]
Oh wow. That's some excellent investigative work @eush77! :D
